### ISSUE TITLE
api-gateway integration test fails on oauth2 client creation

### ISSUE THREAD

**Issue by strekm** [kyma-project/kyma#7021]
**Description**

It happens on 2nd run of api-gateway integration test. First run creates hydra oauth2client, a secret is created also (client name and secret name are hardcoded). At some point test fails (the reason is irrelevant at this point). On 2nd run test attempts to create hydra oauth2client but it is not valid from hydra-maester point of view:

```
2020-01-27T12:46:24.236Z	ERROR	controllers.OAuth2Client	error processing client api-gateway-tests-client/api-gateway-tests 	{"oauth2client": "register", "error": "ID provided in secret api-gateway-tests-secret/api-gateway-tests is assigned to another resource", "errorVerbose": "ID provided in secret api-gateway-tests-secret/api-gateway-tests is assigned to another resource\ngithub.com/ory/hydra-maester/controllers.(*OAuth2ClientReconciler).Reconcile\n\t/go/src/github.com/ory/hydra-maester/controllers/oauth2client_controller.go:164\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.2/pkg/internal/controller/controller.go:216\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.2/pkg/internal/controller/controller.go:192\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.2/pkg/internal/controller/controller.go:171\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357"}
```
As a consequence client is not created in hydra.

**Expected result**

Hydra oauth2client is created and fetching token is possible.

**Actual result**

Hydra doesn't have client registered, fetching token is not possible, test fails

**Steps to reproduce**

**Troubleshooting**

**Comment by piotrmsc** [comment]
Consider waiting  until NS is not deleted fully: https://github.com/kyma-project/kyma/blob/master/tests/integration/api-gateway/gateway-tests/main_test.go#L132

**Comment by jcawley5** [comment]
I have seen the same thing occur when creating a oauth2client via kubectl.  Have to restart the ory-hydra-maester pod for the clients to register.

**Comment by piotrmsc** [comment]
@jcawley5 could you describe steps you did via kubectl and results you had?

**Comment by jcawley5** [comment]
After looking at the log @strekm added this may be a different issue.  I have seen this occur on two fresh installs using version 1.9. using the tutorial at 

https://kyma-project.io/docs/components/api-gateway-v2/#tutorials-expose-and-secure-a-service-register-an-o-auth2-client-and-get-tokens

After submitting the yaml, hydra is not creating the client and no secret will be defined if expecting it to be auto generated.  Nothing shows in the hydra logs when this occurs, but after restarting the pod I haven't experienced the issue again.  On restart I see it pick up any of the clients I defined.

Please let me know if I should create a new issue.

**Comment by Demonsthere** [comment]
With #7266 merged we now have removed the possibility of doubling-down on a client secret, however this does not seem to be the real issue here. 

In hydra we can see the following errors:
```
time="2020-02-21T07:10:04Z" level=error msg="An error occurred" debug="Not Found" description="Client authentication failed (e.g., unknown client, no client authentication included, or unsupported authentication method)" error=invalid_client
time="2020-02-21T07:10:04Z" level=error msg="An error occurred while handling a request" code=404 debug= details="map[]" error="Not Found" reason= request-id=71e3e6f9-1947-4fb1-802e-f9f0037f4633 status="Not Found" 
```

Further investigation is required

**Comment by kubadz** [comment]
Currently, the test fails on Client authentication. Logs from the test:
```
main_test.go:517: Error during Token retrival: oauth2: cannot fetch token: 401 Unauthorized
Response: {"error":"invalid_client","error_description":"Client authentication failed (e.g., unknown client, no client authentication included, or unsupported authentication method)","status_code":401}
```
The test is failing also on the first attempt.

One of failed jobs: https://status.build.kyma-project.io/view/gcs/kyma-prow-logs/logs/post-master-kyma-gke-integration/1234474671542374400

**Comment by piotrmsc** [comment]
We have changed fixed names to generated one for all created resources (namespaces, deployments, clients and so on), added retries while fetching tokens plus configured oauth2 client to always send token in headers and  wait after applying oauth2 client CR before starting test.  The mentioned error did not occur for a week, I will close it. If it happens again, reproducibly please reopen.
